### DISCUSSION TITLE
Camera animation

### DISCUSSION THREAD

**Discussion by zrksyd** [ScanMountGoat/ssbh_editor#140] [Ideas]
I was taking a look at some camera animation files and it seems that it includes near_clip, FOV, far_clip. I think the first step to implementing camera animation would be to have the NUANMB override those values. I included Kazuya's camera animations for you to look through.
[demon_camera.zip](https://github.com/ScanMountGoat/ssbh_editor/files/10503618/demon_camera.zip)

**Comment by ScanMountGoat** [comment]
Getting the values from the nuanmb is the easy part. The harder part is that ssbh_wgpu, the camera editor, and the nuanmb files all use different representations for the camera transformations. The nuanmb files use transform tracks for the camera, but they work a little differently than bone transforms. I already have test data from some in game testing for Cross Mod from a while ago. I don't expect users to want to edit quaternion values, so I'll need to experiment with decomposing the camera transforms into euler angles for display in the camera editor.

**Reply by zrksyd** [reply]
Well if initially you only support viewing camera animations and not editing them, the Euler angle thing does not matter unless you want the user to see the rotation on each frame?

**Reply by ScanMountGoat** [reply]
I have camera animations implemented in ssbh_wgpu. An example would be `cargo run -p ssbh_wgpu_viewer -- /fighter/mario/model/body/c00 --anim /fighter/mario/motion/body/c00/j00win1.nuanmb --camera-anim /camera/mario/c00/j00win1.nuanmb`. It should be straightforward to add support in SSBH Editor at some point. It probably doesn't need its own tab and can be added as an additional drop down in the animations tab.

**Comment by Lilothestitch16** [comment]
Is this implemented for stages that use Camera Animation like Skyloft, Wuhu Island, Mario Circuit Wii U?

**Reply by ScanMountGoat** [reply]
Stage camera animations will be supported too. They just name the animation groups and nodes a little differently.

**Comment by BeyondSmash** [comment]
Hi SMG, to refrain from me making a potentially unnecessary standalone duplicate post - would it be possible to add a rotating camera function to show off/preview a character mod with a parameter setting for "speed of camera rotation" and "height of camera position"? I would love to add some SSBH previews for my portfolio or mod showcase previews.

**Reply by ScanMountGoat** [reply]
This should be possible by just creating a `turntable.nuanmb` using either ssbh_data_py or the Blender addon. If this ends up being difficult to do or something that people use a lot, I'll consider adding a dedicated function in ssbh_editor. I'd like to avoid adding functionality to the program that could be easily handled using another feature like camera anim viewing if possible.

**Reply by BeyondSmash** [reply]
Oh yeah you're right and yeah that makes sense, thank you.

**Comment by ScanMountGoat** [comment]
Camera animations have been implemented and will be included in the next release (0.9.0).
